### PR TITLE
Track length estimator for decay rate tally and other fixes

### PR DESCRIPTION
**Pull request by samuelshaner**
This PR adds the capability to use the track length estimator for the decay-rate tally. Since the decay-rate tally with an analog estimator is only tallied to when there is a delayed neutron created, there will be very few analog tallies to decay-rate. Additionally, a number of other items were addressed:

* `Material.get_nuclide_atom_densities()` method was not properly outputting the nuclide atom densities.
* `mgxs.Library` was not properly adding delayed groups to MDGXS objects.
* Added a `mgxs.MGXS.get_homogenized_mgxs()` method to homogenize MGXS objects. This has been useful for computing the MGXS for the C5G7 benchmark using the given isotopic data. The benchmark homogenizes the fuel, inner gap, inner clad, outer gap, and outer clad, so it is nice to have the functionality to homogenize MGXS within the MGXS framework.
* Fixed ordering of DG in `XSdata._xs_shapes`.
* Added `XSdata.set_inverse_velocity_mgxs()` method.
* Added option to set path of `tallies.xml` file, consistent with the other xml files.
* Updated python example input files to avoid explicitly creating nuclide and element objects.

### CODE REVIEW

File: examples/python/basic/build-xml.py [hunk 1]
```diff
@@ -15,21 +15,16 @@
...
-h1 = openmc.Nuclide('H1')
-o16 = openmc.Nuclide('O16')
-u235 = openmc.Nuclide('U235')
-
 # Instantiate some Materials and register the appropriate Nuclides
 moderator = openmc.Material(material_id=41, name='moderator')
 moderator.set_density('g/cc', 1.0)
-moderator.add_nuclide(h1, 2.)
-moderator.add_nuclide(o16, 1.)
+moderator.add_element('H', 2.)
```
Comment (paulromano) [inline examples/python/basic/build-xml.py]: It's always been a curiosity of mine which is more wrong: Putting only H1 in water? Or putting both H1 and H2 in water but only using a S(a,b) table for H1 (effectively what this does)?

File: examples/python/pincell/build-xml.py [hunk 2]
```diff
@@ -14,86 +14,29 @@
...
-zircaloy.add_nuclide(sn119, 4.1387e-5)
-zircaloy.add_nuclide(sn120, 1.5697e-4)
-zircaloy.add_nuclide(sn122, 2.2308e-5)
-zircaloy.add_nuclide(sn124, 2.7897e-5)
-
-borated_water = openmc.Material(material_id=4, name='Borated water at 975 ppm')
+zircaloy.add_element('Sn', 0.014  , 'wo')
+zircaloy.add_element('Fe', 0.00165, 'wo')
+zircaloy.add_element('Cr', 0.001  , 'wo')
+zircaloy.add_element('Zr', 0.98335, 'wo')
```
Comment (paulromano) [inline examples/python/pincell/build-xml.py]: A wee bit simpler!

File: openmc/material.py [hunk 3]
```diff
@@ -251,7 +251,7 @@ def set_density(self, units, density=None):
 
         Parameters
         ----------
-        units : {'g/cm3', 'g/cc', 'km/cm3', 'atom/b-cm', 'atom/cm3', 'sum', 'macro'}
+        units : {'g/cm3', 'g/cc', 'kg/cm3', 'atom/b-cm', 'atom/cm3', 'sum', 'macro'}
```
Comment (paulromano) [inline openmc/material.py]: What, you don't like your densities in units of km/cm3?!

File: openmc/material.py [hunk 4]
```diff
@@ -641,37 +641,45 @@ def get_nuclide_atom_densities(self):
...
         if not density_in_atom:
-            sum_percent = 0.
-            for n, nuc in enumerate(nucs):
-                x = nuc_densities[n]
-                sum_percent += x * awrs[n]
-            sum_percent = 1. / sum_percent
-            density = -density * sum_percent * \
-                openmc.data.AVOGADRO / openmc.data.NEUTRON_MASS * 1.E-24
+            density = -density / self.average_molar_mass * 1.E-24 \
+                      * openmc.data.AVOGADRO
```
Comment (paulromano) [inline openmc/material.py]: Can you briefly describe what was wrong here?

File: openmc/mgxs_library.py [hunk 5]
```diff
@@ -1644,6 +1644,54 @@ def set_multiplicity_matrix_mgxs(self, nuscatter, scatter=None,
         self._multiplicity_matrix[i] = \
             np.nan_to_num(self._multiplicity_matrix[i])
 
+    def set_inverse_velocity_mgxs(self, inverse_velocity, temperature=294.,
+                                  nuclide='total', xs_type='macro',
+                                  subdomain=None):
+        """This method allows for an openmc.mgxs.InverseVelocity
+        to be used to set the absorption cross section for this XSdata object.
```
Comment (paulromano) [inline openmc/mgxs_library.py]: to set the inverse velocity?

File: tests/test_mgxs_library_condense/results_true.dat [hunk 6]
```diff
@@ -64,12 +64,12 @@
 4     10000             5         1   total  0.001231   0.000105
 5     10000             6         1   total  0.000512   0.000044
    material  delayedgroup  group in nuclide      mean  std. dev.
-0     10000             1         1   total  0.000000   0.000000
-1     10000             2         1   total  0.032739   0.028454
-2     10000             3         1   total  0.120780   0.170809
-3     10000             4         1   total  0.302780   0.109110
-4     10000             5         1   total  0.000000   0.000000
-5     10000             6         1   total  0.000000   0.000000
```
Comment (paulromano) [inline tests/test_mgxs_library_condense/results_true.dat]: Were some of these zeros just due to it being analog before?

File: tests/test_mgxs_library_condense/inputs_true.dat [hunk 7]
```diff
@@ -396,15 +396,15 @@
         <filter bins="0.0 0.625 20000000.0" type="energy" />
         <nuclides>total</nuclides>
         <scores>delayed-nu-fission</scores>
-        <estimator>analog</estimator>
+        <estimator>tracklength</estimator>
```
Comment (wbinventor) [inline tests/test_mgxs_library_condense/inputs_true.dat]: Can you confirm which `MGXS` subclasses are now using tracklength estimators?

File: openmc/material.py [hunk 8]
```diff
@@ -641,37 +641,45 @@ def get_nuclide_atom_densities(self):
...
         density_in_atom = density > 0.
         sum_percent = 0.
 
         awrs = []
-        for n, nuclide in enumerate(nuclides.items()):
+        for nuclide in nuclides.items():
             awr = openmc.data.atomic_mass(nuclide[0])
             if awr is not None:
-                awrs.append(awr / openmc.data.NEUTRON_MASS)
+                awrs.append(awr)
```
Comment (nelsonag) [inline openmc/material.py]: Why did you remove the division by the neutron mass?   `openmc.data.atomic_mass` returns the mass in units of amu, but the awr is the atomic weight ratio, not amu.

File: openmc/mgxs/library.py [hunk 9]
```diff
@@ -524,6 +524,15 @@ def add_to_tallies_file(self, tallies_file, merge=True):
         for domain in self.domains:
             for mgxs_type in self.mgxs_types:
                 mgxs = self.get_mgxs(domain, mgxs_type)
+
+                if mgxs_type in openmc.mgxs.MDGXS_TYPES:
+                    if self.num_delayed_groups == 0:
+                        mgxs.delayed_groups = None
+                    else:
+                        delayed_groups \
```
Comment (nelsonag) [inline openmc/mgxs/library.py]: dont forget a space between the arguments [`range(1, self.num_delayed_groups + 1)`]

File: openmc/mgxs_library.py [hunk 10]
```diff
@@ -296,11 +296,11 @@ def xs_shapes(self):
             self._xs_shapes["[G][G']"] = (self.energy_groups.num_groups,
                                           self.energy_groups.num_groups)
             self._xs_shapes["[DG]"] = (self.num_delayed_groups,)
-            self._xs_shapes["[G][DG]"] = (self.energy_groups.num_groups,
+            self._xs_shapes["[DG][G]"] = (self.energy_groups.num_groups,
```
Comment (nelsonag) [inline openmc/mgxs_library.py]: The left hand side (`[DG][G]`) is inconsistent with the right hand side for these changes.

File: openmc/mgxs_library.py [hunk 11]
```diff
@@ -15,8 +15,8 @@
 # Supported incoming particle MGXS angular treatment representations
 _REPRESENTATIONS = ['isotropic', 'angle']
 _SCATTER_TYPES = ['tabular', 'legendre', 'histogram']
-_XS_SHAPES = ["[G][G'][Order]", "[G]", "[G']", "[G][G']", "[DG]", "[G][DG]",
-              "[G'][DG]", "[G][G'][DG]"]
+_XS_SHAPES = ["[G][G'][Order]", "[G]", "[G']", "[G][G']", "[DG]", "[DG][G]",
+              "[DG][G']", "[DG][G][G']"]
```
Comment (nelsonag) [inline openmc/mgxs_library.py]: What is the driver for this change?

File: src/tally.F90 [hunk 12]
```diff
@@ -691,12 +691,134 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
...
+
+                    ! Get the delayed group for this bin
+                    d = filt % groups(d_bin)
+
+                    ! Compute the yield for this delayed group
+                    yield = nuclides(p % event_nuclide) &
+                         % nu(E, EMISSION_DELAYED, d)
+
+                    associate (rxn => nuclides(p % event_nuclide) % &
+                         reactions(nuclides(p % event_nuclide) % index_fission(1)))
```
Comment (nelsonag) [inline src/tally.F90]: Why is the argument to index_fission always 1? That mens you are using MT=18 data right? I forget, do we not put the yields on the partial fissions?

File: tests/test_mgxs_library_no_nuclides/results_true.dat [hunk 13]
```diff
@@ -129,18 +129,18 @@
...
-4      10000             3         2   total  0.120780   0.170809
-6      10000             4         2   total  0.302780   0.109110
-8      10000             5         2   total  0.000000   0.000000
-10     10000             6         2   total  0.000000   0.000000
+1      10000             1         1   total  0.013445   0.001084
+3      10000             2         1   total  0.032064   0.002658
+5      10000             3         1   total  0.122136   0.010296
+7      10000             4         1   total  0.315269   0.027175
+9      10000             5         1   total  0.889233   0.079368
+11     10000             6         1   total  2.989404   0.266253
```
Comment (nelsonag) [inline tests/test_mgxs_library_no_nuclides/results_true.dat]: This is a pretty big score to have been missed by using analog estimators. Is that really all thats going on here?

File: openmc/mgxs_library.py (hunk 5 above)
Comment (samuelshaner) [inline openmc/mgxs_library.py]: Yeah, I copied the method to set the absorption xs and forgot to change that.

File: tests/test_mgxs_library_condense/results_true.dat (hunk 6 above)
Comment (samuelshaner) [inline tests/test_mgxs_library_condense/results_true.dat]: Yep.

File: openmc/mgxs/library.py (hunk 9 above)
Comment (samuelshaner) [inline openmc/mgxs/library.py]: Will do. I changed it to the following so it would fit on two lines and be less than 80 characters
    
    mgxs.delayed_groups \
        = list(range(1, self.num_delayed_groups+1))

File: openmc/mgxs/mgxs.py [hunk 14]
```diff
@@ -996,6 +996,102 @@ def get_subdomain_avg_xs(self, subdomains='all'):
         avg_xs.sparse = self.sparse
         return avg_xs
 
+    def _get_homogenized_mgxs(self, other_mgxs, denom_score='flux'):
+        """Construct a homogenized mgxs with other mgxs objects.
```
Comment (samuelshaner) [inline openmc/mgxs/mgxs.py]: sure

File: openmc/mgxs/mgxs.py [hunk 15]
```diff
@@ -996,6 +996,102 @@ def get_subdomain_avg_xs(self, subdomains='all'):
...
+        cv.check_iterable_type('other_mgxs', other_mgxs, openmc.mgxs.MGXS)
+        for mgxs in other_mgxs:
+            if mgxs.rxn_type != self.rxn_type:
+                msg = 'Not able to homogenize two MGXS with different rxn types'
+                raise ValueError(msg)
+
+        # Clone this MGXS to initialize the homogenized version
+        homogenized_mgxs = copy.deepcopy(self)
+        homogenized_mgxs._derived = True
+        name = '{} + '.format(self.domain.name)
```
Comment (samuelshaner) [inline openmc/mgxs/mgxs.py]: I like the `hom(...)` format - I'll switch to that.

File: openmc/mgxs_library.py [hunk 16]
```diff
@@ -1644,6 +1644,54 @@ def set_multiplicity_matrix_mgxs(self, nuscatter, scatter=None,
...
+            Provide the macro or micro cross section in units of cm^-1 or
+            barns. Defaults to 'macro'.
+        subdomain : iterable of int
+            If the MGXS contains a mesh domain type, the subdomain parameter
+            specifies which mesh cell (i.e., [i, j, k] index) to use.
+
+        See also
+        --------
+        openmc.mgxs.Library.create_mg_library()
+        openmc.mgxs.Library.get_xsdata
```
Comment (samuelshaner) [inline openmc/mgxs_library.py]: It should have parentheses. This is used for all the setter methods, so I'll change it throughout.

File: openmc/material.py (hunk 8 above)
Comment (samuelshaner) [inline openmc/material.py]: I ended up removing the awrs list altogether. Since the `average_molar_mass` is in amu, I believe we want the nuclide atomic masses in amu.

File: tests/test_mgxs_library_no_nuclides/results_true.dat (hunk 13 above)
Comment (samuelshaner) [inline tests/test_mgxs_library_no_nuclides/results_true.dat]: Yeah, you have to remember we are tallying nu-fission weighted by the decay constant, so the score for group six (which has the largest decay constant) will be quite large.

File: openmc/material.py (hunk 4 above)
Comment (samuelshaner) [inline openmc/material.py]: This is converting the mass density (g/cm3) to atom density (atom/b-cm). First, the mass density (in g/cm3) is divided by the material average molar mass to get the molar density (moles/cm3). Then multiply by avogadro's number to get atom density (atoms/cm3). Finally multiply by 1e-24 to get density in units of atoms/barn-cm. I added a comment saying that we are converting mass density to atom density.

File: openmc/mgxs_library.py (hunk 11 above)
Comment (samuelshaner) [inline openmc/mgxs_library.py]: This is the format used in `src/mgxs_header.F90` so I was just fixing this.

File: openmc/mgxs_library.py (hunk 10 above)
Comment (samuelshaner) [inline openmc/mgxs_library.py]: Fixed.

**Comment by samuelshaner** [comment]
I think all of the comments have been addressed.

**Comment by nelsonag** [comment]
My review is satisfied (How do you mark it as done???)

File: openmc/mgxs/mdgxs.py [hunk 17]
```diff
@@ -1018,6 +1015,34 @@ def xs_tally(self):
 
         return self._xs_tally
 
+    def get_homogenized_mgxs(self, other_mgxs):
+        """Construct a homogenized MGXS with other mgxs objects.
```
Comment (wbinventor) [inline openmc/mgxs/mdgxs.py]: Sorry to be picky, but looks like "...other mgxs objects" should also capitalize "MGXS" to be perfectly explicit.

Comment (samuelshaner) [inline openmc/mgxs/mdgxs.py]: done
